### PR TITLE
allow `_` separators in number literals (stage 1)

### PR DESCRIPTION
**Pull request by momumi**
Implement number literal separators #504

* Underscores `_` may be placed between two digits in a int/float literal
* Consecutive underscores are not allowed
* Fixed parsing bug in exponents of hexadecimal float literals.
  Exponents should always be base 10, but hex characters would be parsed
  inside the exponent and everything after them would be ignored. eg:
  `0x1.0p1ab1` would be parsed as `0x1.0p1`.

### CODE REVIEW

File: test/compile_errors.zig [hunk 1]
```diff
@@ -389,6 +389,110 @@ pub fn addCases(cases: *tests.CompileErrorContext) void {
         "tmp.zig:5:29: error: invalid token: '.'",
     });
 
+    cases.add("invalid underscore placement in float literal - 1",
+        \\fn main() void {
+        \\    var bad: f128 = 0._0;
+        \\})
```
Comment (mikdusan) [inline test/compile_errors.zig]: - please remove invalid trailing token `)` from new tests
- add some compiler-error tests for integer literals ending in underscore:
    - `1000_`
    - `0x1000_`
    - `0b1111_`
    - `0o7777_`

**Comment by momumi** [comment]
I added the changes to the self-hosted as well now.  Let me know if there's anything else I need to implement for this feature.

**Comment by andrewrk** [comment]
Thanks! I'll take a look.

File: src-self-hosted/ir.zig [hunk 2]
```diff
@@ -1311,13 +1311,16 @@ pub const Builder = struct {
         var base: u8 = undefined;
         var rest: []const u8 = undefined;
         if (int_token.len >= 3 and int_token[0] == '0') {
+            rest = int_token[2..];
             base = switch (int_token[1]) {
                 'b' => 2,
                 'o' => 8,
                 'x' => 16,
-                else => unreachable,
+                else => {
```
Comment (LemonBoy) [inline src-self-hosted/ir.zig]: This doesn't even compile, the `else` prong has type `void` as the block doesn't return anything while `base` expects this to be a `u8`.

Comment (andrewrk) [reply src-self-hosted/ir.zig]: This file in particular I plan to delete and start over on.

**Comment by andrewrk** [comment]
Nice work!

**Pull request merged**